c++/src/kj/compat/http.h

```diff
@@ -643,6 +643,7 @@ class WebSocket {
 
   virtual uint64_t sentByteCount() = 0;
   virtual uint64_t receivedByteCount() = 0;
+  virtual kj::Maybe<kj::String> getPreferredExtensions() { return nullptr; }
```

Comment:
This needs some doc comments:

> If pumpTo() / tryPumpFrom() is able to be optimized only if the other WebSocket is using certain extensions (e.g. compression settings), then this method returns what those extensions are. For example, matching extensions between standard WebSockets allows pumping to be implemented by pumping raw bytes between network connections, without reading individual frames.
>
> A null return value indicates that there is no preference. A non-null return value containing an empty string indicates a preference for no extensions to be applied.

And now I see the point I think you were trying to make about distinguishing the outgoing and incoming directions. Technically we could end up pumping between two client-side or two server-side WebSockets. Those cases actually already can't support optimized pumping anyway, at least at present, due to the masking requirements that differ in the two directions.

Perhaps we should add to the comment:

> The extensions are rendered assuming this WebSocket represents the client side of a connection to a remote server. A proxy service could then use this string to fill in the `Sec-WebSocket-Extensions` header of a response to its own client. Calling this method on the server side of a WebSocket doesn't make sense and will not return a useful result.

This is a weird wart, though. I wonder if we can do better by either:
a. Have `getPreferredExtensions()` take a parameter indicating whether the extensions should be rendered for the client->server or server->client direction.
b. Have `getPreferredExtensions()` return a structured representation of the extensions that refers to inbound/outbound rather than client/server, and leave it up to the caller to stringify this appropriately for the context.

Thougths?

Reply:
Yeah, I was thinking it would probably be easier for the caller to specify whether it wanted `client->server` or `server->client`. This function has access to other functions that already stringify compression parameters so I'd rather use those as they've been tested.